library/psa_crypto.c

```diff
@@ -5268,25 +5300,51 @@ static psa_status_t psa_tls12_prf_psk_to_ms_set_key(
     size_t data_length )
 {
     psa_status_t status;
-    uint8_t pms[ 4 + 2 * PSA_TLS12_PSK_TO_MS_PSK_MAX_SIZE ];
+    uint8_t pms[ 4 + PSA_TLS12_PSK_TO_MS_OTHER_SECRET_MAX_SIZE +
+                 PSA_TLS12_PSK_TO_MS_PSK_MAX_SIZE ];
     uint8_t *cur = pms;
 
     if( data_length > PSA_TLS12_PSK_TO_MS_PSK_MAX_SIZE )
         return( PSA_ERROR_INVALID_ARGUMENT );
 
-    /* Quoting RFC 4279, Section 2:
+    /* pure-PSK:
+     * Quoting RFC 4279, Section 2:
      *
      * The premaster secret is formed as follows: if the PSK is N octets
      * long, concatenate a uint16 with the value N, N zero octets, a second
      * uint16 with the value N, and the PSK itself.
+     *
+     * mixed-PSK:
+     * In a DHE-PSK, RSA-PSK, ECDHE-PSK the premaster secret is formed as
+     * follows: concatenate a uint16 with the length of the output of RSA
+     * key wrapping or of (EC)DHE key agreement, the output itself,
+     * uint16 with the length of PSK, and the PSK itself.
+     * For details please check:
+     * - RFC 4279, Section 4 for the definition of RSA-PSK,
+     * - RFC 4279, Section 3 for the definition of DHE-PSK,
+     * - RFC 5489 for the definition of ECDHE-PSK.
+     *
+     * Salt holds the output of RSA key wrapping or
+     * of (EC)DHE key agreement.
      */
 
+    if ( prf->state == PSA_TLS12_PRF_STATE_SALT_SET )
+    {
+        *cur++ = MBEDTLS_BYTE_1( prf->salt_length );
+        *cur++ = MBEDTLS_BYTE_0( prf->salt_length );
+        memcpy( cur, prf->salt, prf->salt_length );
+        cur += prf->salt_length;
```

Comment:
Actually I think we could clear & free `salt` now, (or before exiting the function, at the same time as we clear the PMS). Doesn't make much of a difference, but as a general rule it's good practice to clear secrets as soon as possible.

Reply:
I agree, but on the other hand it is good free all buffers in the same place (probably other inputs could also be freed earlier).